src/CommunityToolkit.Maui/Views/GravatarImageSource.shared.cs

```diff
@@ -0,0 +1,176 @@
+	public override bool IsEmpty => Uri == null;
+
+	/// <summary>Gets or sets the URI for the image to get.</summary>
+	[System.ComponentModel.TypeConverter(typeof(UriTypeConverter))]
+	public Uri Uri { get; private set; }
+
+	/// <summary>Gets or sets the control size property.</summary>
+	int GravatarSize
+	{
+		get => gravatarSize;
+		set => gravatarSize = Math.Clamp(value, 1, 2048);
+	}
+
+	/// <summary>Returns the Uri as a string.</summary>
+	/// <returns>String of the URI.</returns>
+	public override string ToString()
+	{
+		return $"Uri: {Uri}, Email:{Email}, Size:{GravatarSize}, Image: {DefaultGravatarName(Image)}, CacheValidity:{CacheValidity}, CachingEnabled:{CachingEnabled}";
+	}
+
+	/// <summary>On parent set.</summary>
+	protected override void OnParentSet()
+	{
+		object element = Parent;
+		base.OnParentSet();
+		if (element is Microsoft.Maui.Controls.Image parentImage)
+		{
+			var height = parentImage.Height >= 0 ? parentImage.Height : Math.Max(parentImage.HeightRequest, GravatarImageSourceDefaults.DefaultSize);
+			var width = parentImage.Width >= 0 ? parentImage.Width : Math.Max(parentImage.WidthRequest, GravatarImageSourceDefaults.DefaultSize);
+			var sizeFromParent = (int)Math.Max(width, height);
+			if (GravatarSize != sizeFromParent)
+			{
+				GravatarSize = sizeFromParent;
+				HandleNewUriRequested(Email, Image);
+			}
+		}
+	}
+
+	static string DefaultGravatarName(DefaultImage defaultGravatar)
+			=> defaultGravatar switch
+			{
+				DefaultImage.FileNotFound => "404",
+				DefaultImage.MysteryPerson => "mp",
+				_ => $"{defaultGravatar}".ToLower(),
+			};
+
+	static string GetMd5Hash(string? str)
+	{
+		if (string.IsNullOrWhiteSpace(str))
+		{
+			return string.Empty;
+		}
+
+		using var md5 = MD5.Create();
+		Span<byte> hash = md5.ComputeHash(Encoding.UTF8.GetBytes(str));
+		return BitConverter.ToString(hash.ToArray(), 0, hash.Length).Replace("-", string.Empty, StringComparison.OrdinalIgnoreCase).ToLowerInvariant();
+	}
+
+	static void OnDefaultImagePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested(gravatarImageSource.Email, (DefaultImage)newValue);
+	}
+
+	static void OnEmailPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested((string?)newValue, gravatarImageSource.Image);
+	}
+
+	void HandleNewUriRequested(string? email, DefaultImage image)
+	{
+		Uri = string.IsNullOrWhiteSpace(email)
+			? new Uri($"{defaultGravatarImageAddress}?s={GravatarSize}")
+			: new Uri($"{defaultGravatarImageAddress}{GetMd5Hash(email)}?s={GravatarSize}&d={DefaultGravatarName(image)}");
+
+		OnUriChanged();
+	}
+
+	void OnUriChanged()
+	{
+		CancellationTokenSource?.Cancel();
```

Comment:
This isn't very efficient, because we will throw an `CancelledException` every time that user taps, and probably do a new HTTP request every time. Can add wait some `ms` before calling this method and the `DownloadStreamAsync`?

This will be something like [this](https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui/Behaviors/UserStoppedTypingBehavior.shared.cs)

Reply:
@pictos I was looking to try and make this more efficient as it annoyed me too!  The pointer to a working example is very much appreciated, and I'll look to try and implement that!

Reply:
Implemented.  From my testing it looks like it does reduce/eliminate diplicate calls when binding properties have changed.